### ISSUE TITLE
[ENH]: Generalized error bars

### ISSUE THREAD

**Issue by hjstein** [matplotlib/matplotlib#28614]
### Problem

The name of the errorbar() function suggests that it is specifically for adding error bars to a plot.  However, it's actually much more general than that, giving a very convenient way to add additional data to a plot.  For example, I've been using it to denote 25% and 75% percentiles when plotting means.  The problem is that this more general usage is precluded because in such use cases, the error bars can be negative, and recent versions throw an error in this case.

### Proposed solution

My proposed solution is to get rid of this test.  Users can easily see if their error bars bound the function they're plotting, and may very well, for reasons as noted above, want to use the functionality in cases where the error bars are not positive.

Accepting negative error bars was the behavior of errorbar() up until commit #2cd1846751, which throws an error if the error bars are not positive.  Later changes refined this to throw an error for negative error bars.  I believe this test was incorporated into version 2.6.

My enhancement request is for this test to be removed.  The patch would be:
```
diff --git a/lib/matplotlib/axes/_axes.py b/lib/matplotlib/axes/_axes.py
index b42d526701..f92c6e785d 100644
--- a/lib/matplotlib/axes/_axes.py
+++ b/lib/matplotlib/axes/_axes.py
@@ -3474,8 +3474,6 @@ class Axes(_AxesBase):
               errors.
             - *None*: No errorbar.
 
-            All values must be >= 0.
-
             See :doc:`/gallery/statistics/errorbar_features`
             for an example on the usage of ``xerr`` and ``yerr``.
 
@@ -3744,10 +3742,6 @@ class Axes(_AxesBase):
                     "Use NaN if you want to skip a value.")
 
             res = np.zeros(err.shape, dtype=bool)  # Default in case of nan
-            if np.any(np.less(err, -err, out=res, where=(err == err))):
-                # like err<0, but also works for timedelta and nan.
-                raise ValueError(
-                    f"'{dep_axis}err' must not contain negative values")
             # This is like
             #     elow, ehigh = np.broadcast_to(...)
             #     return dep - elow * ~lolims, dep + ehigh * ~uplims
````

**Comment by jklymak** [comment]
I'm not quite following. If y is 10 and you want the errorbar to go from 7 to 12 you put (3, 2) in for the y error.  If you want it to go from -5 to 15 you use (15, 5).  If you mean you want the errorbar to go from 7 to 9 then I agree this won't work, but then it's not an errorbar.  If I were to generalize I'd maybe allow (3, None) and only draw the lower part of the error bar.  But otherwise I'm pretty confused about what your graphs look like with the error bar not centred on the data point.

**Comment by timhoffm** [comment]
Please check the discussion in #26187 to understand why we're not supporting negative error values. If an equivalent functionality comes, it should rather be through a `yrange` parameter (#26220).

**Comment by rcomer** [comment]
Hi @hjstein I guess you are currently doing something like

```python
import matplotlib.pyplot as plt
import numpy as np

means = np.array([3, 6, 4])
quartiles = np.array([[2, 4], [4, 7], [1, 7]])

fig, ax = plt.subplots()

# Define quartiles as "errors" relative to mean as that's what errorbar
# function requires.
low_err = means - quartiles[:, 0]
high_err = quartiles[:, 1] - means

# Plot means and interquartile ranges.
ax.errorbar(range(3), means, yerr=[low_err, high_err], fmt='o')
```
But you can get the same result by plotting the means and interquartile ranges separately - setting `fmt='none'` in `errorbar` tells it to skip plotting the central y-values:

```python
fig, ax = plt.subplots()

# Find the midpoints of the interquartile ranges.
mids = np.mean(quartiles, axis=1)
# Define the half-ranges as "errors" relative to the midpoint as that's what
# errorbar function requires.
err = quartiles[:, 1] - mids

# Plot the interquartile ranges.
c = ax.errorbar(range(3), mids, yerr=err, fmt='none')

# Plot the means.
ax.plot(means, 'o')
```
Although this requires an extra line, I would argue that it's clearer code as it separates concerns between plotting means and quartiles (which are obviously not related in the way `errorbar` assumes).  It also works fine if the mean is not in the interquartile range.

**Comment by jklymak** [comment]
I prefer @rcomer method to a conflicting range parameter 

Note that one can also make error bars using arrows: https://matplotlib.org/stable/users/explain/text/annotations.html#customizing-annotation-arrows

**Comment by hjstein** [comment]
Thank you everyone for all your comments, and for pointing me to the previous discussion (namely https://github.com/matplotlib/matplotlib/issues/26187). I have to apologize for not finding & reviewing it before making this suggestion.

To give context, as requested by  @jklymak in https://github.com/matplotlib/matplotlib/issues/26187#issuecomment-1607847024 and @oscargus in https://github.com/matplotlib/matplotlib/issues/26187#issuecomment-1609500189, the way I see it, the issue is the difference between `boxplot()` and `errorbar()`.  My view is that there are two fundamental differences.  Firstly, `boxplot()` takes the raw data as input and visually summarizes it.  The `errorbar()` function instead takes the summarized data as input and displays it.  Secondly, `boxplot()` is designed for a categorical, discrete X axis.  The `errorbar()` function can handle a continuous X axis.

So, the general use case is to use `errorbar()` to get a continuous version of `boxplot()` that takes the summarized data as input instead of the full data set.  To do this, my code looks like:
```
    p1 = ax.errorbar(sds.index, sds["mean"], yerr=[sds["mean"] - sds["min"], sds["max"] - sds["mean"]])
    p2 = ax.errorbar(sds.index, sds["mean"], yerr=[sds["mean"] - sds["25%"], sds["75%"] - sds["mean"]], marker=".")
    plt.fill_between(sds.index, sds["mean"] - sds["std"], sds["mean"] + sds["std"], alpha=0.3)
```
X is continuous, so I can't use `boxplot()`.  Also, I need the stats for other purposes, and it's a lot of data, so I'm calculating them separately & after calculating it, I want to plot it.  Hence `errorbar()` instead of `boxplot()`.

My issue is that I did this a while ago, before negative `yerr` values was an error, and when I tried rerunning it, my code was failing because of the negative `yerr` error.  Thinking I had inadvertently broke something in the interim, I had to dig into old code & old data to see what was going on and why it was failing now, only to find that it was due to the test for negative `yerr` values that was added in version 3.6.

I realize I can modify my code in various ways to get around this error, but I don't see why users should have to think about & make such modifications.  If there's a downstream reason in the implementation that allowing negative `yerr` values through would cause substantial complications, then I could accept it.  Otherwise, it's just making it harder for me to do what I want.

@timhoffm suggests in https://github.com/matplotlib/matplotlib/issues/26220 to instead add a `yrange` argument.  I agree that it would be a convenient addition, allowing me to avoid the above subtraction, but I'm not a big fan of this addition.  My problem is that you then have two ways of specifying the same error bars - `yerr` & `yrange`.  I think it adds API complexity without buying additional functionality.  And then, what if `yrange=(a, b)`, and `y` is not between `a` & `b`?  If negative error bars signal an error, then shouldn't `a > y` or `b < y` signal an error as well?  If you maintain consistency, then it doesn't solve my problem.

@timhoffm in https://github.com/matplotlib/matplotlib/issues/26187#issuecomment-1607592431) notes that negative yerrs are ambiguous.  He gives the example of:
```
   errorbar(x, y=[0, 0, 0], yerr=[(-2, 10), (-3, 10), (-4, 10)])
```
noting that a user could think the first error bar should go from -2 to 10 rather than from 2 to 10.  Like @anntzer in https://github.com/matplotlib/matplotlib/issues/26187#issuecomment-1615840825, I don't find it ambiguous.  But I think the documentation can be changed so as to reduce ambiguity.  Currently, it has phrases like "Separate - and + values for each bar", which can be interpreted in different ways.  Maybe the documentation could be more explicit, as in something along the lines of:
 - scalar and shape(N): Error bars are symmetric, ranging from `x - xerr` to `x + xerr`, and `y - yerr` to `y + yerr`, respectively.
 - shape(2,N): Error bars are asymmetric, ranging from `x - xerr[0]` to `x + xerr[1]`, and `y - yerr[0]` to `y + yerr[1]`, respectively.

I know this doesn't cover the `[(a0, b0), (a1, b1), ...]` case, but maybe something along these lines could be written that would help to reduce ambiguity.

As a user, I certainly agree that clear, unambiguous APIs and documentation should be aspired to.  But, in this particular case, if I misunderstand and use negatives as in the above example, it will be obvious when I see the graph that I misunderstood, and I can easily fix it.  So I don't find the ambiguity noted here as a strong reason for throwing an error when yerr is negative.  At best, maybe a warning instead of throwing an error, if, in fact, it happens to be the case that it keeps coming up as a question from confused users.

To summarize this rather long comment, I think in general one should think of `errorbar()` as a simplified version of `boxplot()` that works on a continuous X axis & uses summarized data.  I don't think users will be confused by the bars not bracketing the `y` values, as they can plainly see whether or not that's what they want.  Throwing an error for negative `yerr` values then requires users to work around this restriction.  If allowing negative `yerr` values would substantially complicate the implementation of `errorbar()`, then I would accept testing for it and throwing an error rather than failing in some obscure way.  But, given that negative `yerr` values used to be accepted, I don't think that this is the case, so I would much prefer that the test for negative `yerr` values be removed and the documentation to be updated appropriately.

**Comment by rcomer** [comment]
You can make boxplots with pre-summarized data using [bxp](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.bxp.html#matplotlib.axes.Axes.bxp).  I'm not sure I follow the argument about the continuous x-axis, but if it's just that you want to use `sds.index` to controls the x locations, there is the _positions_ parameter in `bxp`.

**Comment by jklymak** [comment]
@hjstein unfortunately too many people were confused by negative limits that we had to lock that out, certainly far more than want to have error bars that are not centred on the data point.  I don't think there is any appetite to revert that check.

**Comment by hjstein** [comment]
> You can make boxplots with pre-summarized data using [bxp](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.bxp.html#matplotlib.axes.Axes.bxp). I'm not sure I follow the argument about the continuous x-axis, but if it's just that you want to use `sds.index` to controls the x locations, there is the _positions_ parameter in `bxp`.

Thanks.  I didn't know about [bxp](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.bxp.html#matplotlib.axes.Axes.bxp).  If I could also get bxp to do lines instead boxes, that would almost be what I need.  The remaining item is a continuous X axis, meaning connecting the $(x,y)$ points, and shading continuously instead of using discrete boxes, then that would do what I need.  In lieu of that, I'll need to continue to work with `errorbar()` and work around it throwing an error for negative `yerr` values.

**Comment by hjstein** [comment]
> @hjstein unfortunately too many people were confused by negative limits that we had to lock that out, certainly far more than want to have error bars that are not centred on the data point. I don't think there is any appetite to revert that check.

Sorry to hear that.  If that's what was going on, I could understand locking it out.  But that could also be a documentation issue.  Clarifying the documentation might prevent that as well, without locking people out of that functionality.  As would getting an explanatory warning instead of throwing an error.

**Comment by hjstein** [comment]
I tried @rcomer's suggestion above of modifying the code and using `fmt='none'`, but I'm having trouble getting it to work.  The problem is replicating the previous graph's appearance.  Here's the test data:
```
df = pd.DataFrame()
df["X"] = [-2, -1, 0, 1, 2]
df["mean"] = df["X"]**2
df["min"] = df["mean"]/2
df["max"] = df["mean"]*2
df["25%"] = df["mean"] - 0.5
df["50%"] = df["mean"] - 0.4
df["75%"] = df["mean"] - 0.25
df["std"] = df["mean"]*0.25
df = df.set_index("X")
```
Here's the code, which works prior to version 3.6:
```
fontsize=20
with plt.rc_context({'font.size': fontsize}):
    _, ax = plt.subplots(figsize=(16,8))
    p1 = ax.errorbar(df.index, df["mean"], yerr=[df["mean"] - df["min"], df["max"] - df["mean"]])
    p2 = ax.errorbar(df.index, df["mean"], yerr=[df["mean"] - df["25%"], df["75%"] - df["mean"]], marker=".")
    plt.fill_between(df.index, df["mean"] - df["std"], df["mean"] + df["std"], alpha=0.3)

    plt.legend([p2, p1], ["Mean SD", "Extrema"], loc="upper left")
    plt.show()
```
This produces the attached OldGraph.

The attempt at reproducing this:
```
fontsize=20
with plt.rc_context({'font.size': fontsize}):
    _, ax = plt.subplots(figsize=(16,8))
    p1 = ax.errorbar(df.index, df["mean"], yerr=[df["mean"] - df["min"], df["max"] - df["mean"]], marker=".", ecolor='#1f77b4', color='#ff7f0e')
    p2 = ax.errorbar(df.index, df["50%"], yerr=[df["50%"] - df["25%"], df["75%"] - df["50%"]], ecolor='#ff7f0e', color='#ff7f0e', fmt="none")
    plt.fill_between(df.index, df["mean"] - df["std"], df["mean"] + df["std"], alpha=0.3)

    plt.legend([p2, p1], ["Mean SD", "Extrema"], loc="upper left")
    plt.show()
```
This produces the attached NewGraph.

To get the 25% and 50% on top of the max and min error bars, I need to graph them after the first.  In the old graph, the color of the mean line was from the second graph, but I have to turn off the line in the 2nd graph because it's now the 50% line, so the mean line color is now wrong.  Moreover, if I don't specify colors, the 2nd call to `errorbar()`, instead of using the next color in the cycle, uses the color of the line from the first call, so the overlapping lines are invisible.

To fix that, I need to explicitly set all of the colors, which is a pain, but doable.  However, what's worse is that because I'm skipping the `y` line in the 2nd call, and because I'm using different colors in the first to get the right color in the mean line, now the legend is different.  It's not terrible, but I need to update the code to work with the newer version, and these graphs go into a report, so I can't have changes.

The bottom line is that it's actually not so easy to replicate the old behavior.  Can someone help me replicate the old graph?

(FYI - I used "click to add files" to add the graphs, but I didn't see them when looking at the preview, so I'm not sure if they came through.)

**Comment by hjstein** [comment]
Looks like I can't attach files or include images.  If someone can share them, I'd be happy to email them over.

**Comment by rcomer** [comment]
I think this does the trick:

```python
with plt.rc_context({'font.size': fontsize}):
    _, ax = plt.subplots(figsize=(16,8))
    p1 = ax.errorbar(df.index, df["mean"], yerr=[df["mean"] - df["min"], df["max"] - df["mean"]], marker=".", color='C0')
    p2 = ax.errorbar(df.index, df["50%"], yerr=[df["50%"] - df["25%"], df["75%"] - df["50%"]], color='C1', fmt="none")
    l, = ax.plot(df.index, df["mean"], color='C1', zorder=5, marker='.')
    plt.fill_between(df.index, df["mean"] - df["std"], df["mean"] + df["std"], alpha=0.3)

    plt.legend([(p2, l), p1], ["Mean SD", "Extrema"], loc="upper left")
```
Note
- "C0", "C1", etc get you the colours from the colour cycle.  [Specifying Colors](https://matplotlib.org/stable/users/explain/colors/colors.html#sphx-glr-users-explain-colors-colors-py)
- Setting _zorder_ to some high number to get the mean line on top in `plot`.  [Zorder Demo](https://matplotlib.org/stable/gallery/misc/zorder_demo.html)
- Using a tuple of artists for the first handle in `legend`.  [Legend Guide](https://matplotlib.org/stable/users/explain/axes/legend_guide.html#legend-handlers)

**Comment by hjstein** [comment]
Yup, @rcomer.  You got it.  Thanks.  Tricky, but doable, which is part of my point...
